File: src/main/java/org/willxu/algorithm/service/intlist/BinaryTreeInorderTraversal.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.intlist;

import org.willxu.algorithm.provide.TreeNode;

import java.util.List;

public interface BinaryTreeInorderTraversal {
    /**
     * Given the root of a binary tree, return the inorder traversal of
     * its nodes' values.
     *
     * Constraints:
     * -100 <= Node.val <= 100
     *
     * Follow up:
     * Recursive solution is trivial, could you do it iteratively?
     *
     * @param root The number of nodes in the tree is in the range
     *            [0, 100].
     * @return the inorder traversal of its nodes' values.
     */
    List<Integer> inorderTraversal(TreeNode root);
}

File: src/leetcode/p0000_0099/p0094_binary_tree_inorder_traversal/binary_tree_inorder_traversal.py
from abc import ABC, abstractmethod
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTreeInorderTraversal(ABC):
    """
    Given the root of a binary tree, return the inorder traversal of its nodes' values.
    """
    @abstractmethod
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        """
        Args:
            root:
                - The numver of nodes in the tree is in the range [0, 100]
                - -100 <= Node.val <= 100
        """
        pass

class BinaryTreeInorderTraversalRecursive(BinaryTreeInorderTraversal):
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result: List[int] = []
        self.inorder(root, result)
        return result

    def inorder(self, node, result):
        if not node:
            return
        self.inorder(node.left, result)
        result.append(node.val)
        self.inorder(node.right, result)
